### PR TITLE
Fixes 4497 : this a photo of place X: Do not ask if the item (depiction) is already selected for addition and the photo already has a caption

### LINKED ISSUE #4497: Is this a photo of place X: Do not ask if the item (depiction) is already selected for addition and the photo already has a caption
Latest master

In such cases, asking again has no positive benefit, and has some drawbacks:
- Takes time to think and answer
- If "Yes" is tapped, the caption (copied from previous media) is overwritten by a less useful caption

### PR DESCRIPTION
**Pull request by Ayan-10**
**Description (required)**

Fixes #4497 

If users already selected their preference about a nearby place while uploading, asking again has no positive benefit, and has some drawbacks. 
That's why I
1. Stores every nearby place that is found while uploading and corresponding user response in a Hashmap
2. Whenever a nearby place is found, it will check the user response and place availability and act accordingly.

**Tests performed (required)**

Tested latest ProdDebug Realme GT master with API level 30.

### CODE REVIEW

**Review by nicolas-raoul — COMMENTED** [review]
These depictions already stored somewhere, right? (so that they get pre-selected in the depictions selection step).
How about reading from that already implemented storage place, rather than implementing a new storage (HashMap)?

**Comment by Ayan-10** [comment]
@nicolas-raoul Yes, the depictions are pre-selected in a list. But I think we can not use that storage for several reasons. I am mentioning them below.

First of all, with each place/depiction, we need to store the user response (is this a photo of place X yes/no). So I have to link the response with each place/depiction. So I have to use a hashmap to link the response with each place/depiction.

The reason behind this is, suppose a user uploaded 6 images together (multiple uploads). 
2 of having coordinates 37.4221,122.0841 which is linked to a place/depiction
another 2 of them have coordinates 48.8484,2.35926 which is linked to another place/depiction
another 2 of them have coordinates 40.6892,74.0445 which is linked to another place/depiction.

These 6 images are selected haphazardly (Two images linked to the same place are not consecutive while selecting for uploads). 

User selects **yes** for 37.4221,122.0841 place when 1st image of 37.4221,122.0841 coordinates shows is this a photo of place X dialog. Then an image with 48.8484,2.35926 coordinates came after clicking next. User select **no** for 48.8484,2.35926. Then an image with 40.6892,74.0445 coordinates came after clicking next. User select **yes** for 40.6892,74.0445. Now, when three 2nd images with each location will come we have to check the user response (is this a photo of place X yes/no) for the 1st occurrence of images to determine we should show the dialog or not. That is why I have to link user responses with each place.


Now, coming to the 2nd reason. Yes, the depictions are preselected in a list of DepictedItem. In the current situation, the app is getting the place for every upload and converting them into DepictedItem. They are using sitelinks of places to fetch DepictedItem from the server. So with 1 place, there can be more than one depiction. So with the preselected in a list of DepictedItem we can't determine that is this place of upload item occurred for any other previous upload item or not. To determine that question we need place names and place coordinates which are not available in DepictedItem. So, I have to store Place objects either.

Hope I am able to express all the reasons.

Thanks.

**Comment by nicolas-raoul** [comment]
That makes sense, thank you for the explanation! :-)

File: app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java [hunk 1]
```diff
@@ -117,6 +118,12 @@
 
     public static final String EXTRA_FILES = "commons_image_exta";
 
+    /**
+     * Stores all nearby places found and related users response for
+     * each place while uploading media
+     */
+    public static HashMap<Place,Boolean> existingPlaces;
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java:125]: I think we need a more explicit name for this.
How about `nearbyPopupAnswers`?
Or any other idea?

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
It works great so far in my testing :-)

File: app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java (hunk 1 above)
Comment (Ayan-10) [reply app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java:125]: @nicolas-raoul Yes, `nearbyPopupAnswers` will be better. One other name can be `nearbyPopupResponses`. Tell me which is more accurate to you I will change it.

Thanks

Comment (nicolas-raoul) [reply app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java:125]: You are right, I thought about `Responses` first, but we'd better avoid possible ambiguity with _HTTP responses_, I think.

**Review by nicolas-raoul — APPROVED** [review]

**Pull request merged**